src/app/qgsappbrowserproviders.cpp

```diff
@@ -627,38 +629,173 @@ QgsBookmarkManagerItem::QgsBookmarkManagerItem( QgsDataItem *parent, const QStri
-  connect( mManager, &QgsBookmarkManager::bookmarkRemoved, this, [ = ]( const QString & ) { depopulate(); refresh(); } );
+  connect( mManager, &QgsBookmarkManager::bookmarkAdded, this, [ = ]( const QString & id )
+  {
+    QgsBookmark newDetails = mManager->bookmarkById( id );
+    if ( newDetails.group().isEmpty() )
+      addChildItem( new QgsBookmarkItem( this, newDetails.name(), newDetails, mManager ), true );
+    else
+    {
+      if ( QgsBookmarkGroupItem *newGroup = groupItem( newDetails.group() ) )
+      {
+        // existing group, add this bookmark to it
+        newGroup->addBookmark( newDetails );
+      }
+      else
+      {
+        // need to create a new group for this (will automatically add the new bookmark)
+        addChildItem( new QgsBookmarkGroupItem( this, newDetails.group(), mManager ), true );
+      }
+    }
+  } );
+  connect( mManager, &QgsBookmarkManager::bookmarkChanged, this, [ = ]( const QString & id )
+  {
+    QgsBookmark newDetails = mManager->bookmarkById( id );
+
+    // have to do a deep dive to find the old item...!
+    const QVector<QgsDataItem *> c = children();
+    for ( QgsDataItem *i : c )
+    {
+      if ( QgsBookmarkItem *bookmarkItem = qobject_cast< QgsBookmarkItem * >( i ) )
+      {
+        if ( bookmarkItem->bookmark().id() == id )
+        {
+          // found the target! now, what's changed?
+          if ( bookmarkItem->bookmark().group() == newDetails.group() )
+          {
+            // good, not the group. Just update the existing bookmark then.
+            bookmarkItem->setBookmark( newDetails );
+            return;
+          }
+          else
+          {
+            // group has changed, ouch!
+            // first remove from existing group
+            deleteChildItem( bookmarkItem );
+            // and add a child to the new group
+            if ( QgsBookmarkGroupItem *newGroup = groupItem( newDetails.group() ) )
+            {
+              newGroup->addBookmark( newDetails );
+            }
+            else
+            {
+              // need to create a new group for this (will automatically add the new bookmark)
+              addChildItem( new QgsBookmarkGroupItem( this, newDetails.group(), mManager ), true );
+            }
+          }
+          break;
+        }
+      }
+      else if ( QgsBookmarkGroupItem *group = qobject_cast< QgsBookmarkGroupItem * >( i ) )
+      {
+        if ( QgsBookmarkItem *bookmarkItem = group->childItemById( id ) )
+        {
+          // ok, found old group, now compare
+          if ( bookmarkItem->bookmark().group() == newDetails.group() )
+          {
+            // good, not the group. Just update the existing bookmark then.
+            bookmarkItem->setBookmark( newDetails );
+          }
+          else
+          {
+            // group has changed!
+            // first remove from existing group
```

Comment:
I think we need some additional code to remove the departed group if that group didn't have any other bookmarks in it.